mantidproject/mantid
# 16011 biosans wing detector

New additional detector for BioSANS
- Now BioSANS and GP SANS have `detector1` as `rectangular_detector`, which implies a lot of modifications on the getPixelId from pixel (x,y) in the detector
- LoadSpice2D has to work for both BioSANS and GPSANS.
- LoadSpice2D looks for more than one `<Data ...>` tag. BioSANS now has 2.
- ~~LoadSpice2D also translates the detector according to `<detector_trans>`~~.
- LoadSpice2D uses `setStartAndEndTime` to be able to handle multiple IDFs.
- MaskedFullComponent masks a component by name:
  - `MaskComponent(component_name)`
- MaskedComponent is used along with MaskedEdges to mask the edges of a component
  - `Mask(nx_low=0, nx_high=0, ny_low=0, ny_high=0, component_name=None)`
- SANSSolidAngleCorrection applies a different correction for Wing detector:
  - `SolidAngle(detector_tubes=False, detector_wing=False)`

**To test:**

Use these files:
[data.zip](https://github.com/mantidproject/mantid/files/275340/data.zip)

Run two separate scripts. Don't forget to change the path.

The result should be 2 nice curves for Iq and Iqxy, with q ranges ranges different for both of them, as below:

![screenshot from 2016-05-20 08 48 56](https://cloud.githubusercontent.com/assets/1866557/15440237/659517d6-1ea2-11e6-9625-408c109994b3.png)

Scripts:

``` python
# HFIR Bio-SANS reduction script - qians@ornl.gov Shuo Qian 2016 May

import mantid
from mantid.simpleapi import *
from reduction_workflow.instruments.sans.hfir_command_interface import *
#import openpyxl
#from openpyxl import load_workbook
#from openpyxl import Workbook


################Set Bio-SANS .xml data folder and root path:###########################
xml_data_folder="/home/rhf/Documents/SANS/BioSans/20160520-WingWarnings"#Change this directory to your local path containing the .xml datafiles
root_path=xml_data_folder+"/"
###############File name prefix and suffix#########################################

SampleScattering="test_0031.xml"
BgScattering       ="test_bg_0031.xml"
wing_q_range      ="0.05,0.0025,1"
dark60m_curr_file="psBioSANS_exp291_6mDark.xml" #detector at 6m & 1.13 m; 15m &1.13m for dark
flood_field_file     ="psBioSANS_exp291_Flood.xml" #a 6m&1.13 flat field
EmptyTrans         = "test_0031_trans_0001.xml"#6m MT beam file
beam60m_center_file = EmptyTrans
BgTrans             ="test_bg_0031_trans_0001.xml"
main_out_name =SampleScattering+"_main"
wing_out_name = SampleScattering+"_wing"
```

``` python
#Main detector reduction ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
BIOSANS()
#SetSampleDetectorOffset(3)
#SetWavelength(lmbda6, 0.15)
SolidAngle(detector_tubes=True)
DataPath(xml_data_folder)
DarkCurrent(dark60m_curr_file)
MonitorNormalization()
SetAbsoluteScale(1)
IQxQy(nbins=80)
#NoIQxQy()
#for main detector
MaskComponent("wing_detector")
Mask(nx_low=2, nx_high=2, ny_low=20, ny_high=20)
AzimuthalAverage(n_bins=80, n_subpix=1, log_binning=True)
# Mask main detector
#MaskComponent("detector1")
#Mask(nx_low=1, nx_high=1, ny_low=20, ny_high=20, component_name="wing_detector")
#AzimuthalAverage(binning="0.11,0.005,0.5", n_subpix=1,log_binning=False)
DirectBeamCenter(beam60m_center_file)
ThetaDependentTransmission(True)
DirectBeamTransmission(EmptyTrans, EmptyTrans, beam_radius=3)
TransmissionDarkCurrent(dark60m_curr_file)
AppendDataFile(SampleScattering, workspace=main_out_name)
Background(BgScattering)
BckDirectBeamTransmission(BgTrans, EmptyTrans, beam_radius=3)
BckThetaDependentTransmission(True)
BckTransmissionDarkCurrent(dark60m_curr_file)
SaveIq(process='')
Reduce()
```

``` python
#Wing detector reduction    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
BIOSANS()
#SetSampleDetectorOffset(3)
#SetWavelength(lmbda6, 0.15)
SolidAngle(detector_wing=True)
#NoSolidAngle()
DataPath(xml_data_folder)
DarkCurrent(dark60m_curr_file)
MonitorNormalization()
SetAbsoluteScale(.1)
IQxQy(nbins=80)
#NoIQxQy()
#for main detector
#MaskComponent("wing_detector")
#Mask(nx_low=2, nx_high=2, ny_low=20, ny_high=20)
#AzimuthalAverage(n_bins=80, n_subpix=1, log_binning=True)
# Mask main detector
MaskComponent("detector1")
Mask(nx_low=0, nx_high=1, ny_low=20, ny_high=20, component_name="wing_detector")
AzimuthalAverage(binning=wing_q_range, n_subpix=1,log_binning=False)
DirectBeamCenter(beam60m_center_file)
ThetaDependentTransmission(True)
DirectBeamTransmission(EmptyTrans, EmptyTrans, beam_radius=3)
TransmissionDarkCurrent(dark60m_curr_file)
AppendDataFile(SampleScattering, workspace=wing_out_name)
Background(BgScattering)
BckDirectBeamTransmission(BgTrans, EmptyTrans, beam_radius=3)
BckThetaDependentTransmission(True)
BckTransmissionDarkCurrent(dark60m_curr_file)
SaveIq(process='')
Reduce()

```



Fixes #16011 .



---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
##### Functional Tests
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.
